### PR TITLE
New module: proxmox_snap

### PR DESCRIPTION
**Pull request by Thulium-Drake**
This commit adds proxmox inventory module and proxmox_snap for snapshot management

##### SUMMARY

proxmox_snap: based on 'proxmox', adapted to manage snapshots for LXC/Qemu machines (OpenVZ might work, but I didn't test it as it is no longer supported in PVE)

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

proxmox_snap

### CODE REVIEW

**Comment by Thulium-Drake** [comment]
bot_status

**Comment by Thulium-Drake** [comment]
Hmm, are the bot and me looking to the same file? I fixed all doc errors it mentioned in the first post (missing types etc..), but it now mentions even more errors?

What did I do wrong?

**Comment by Thulium-Drake** [comment]
bot_status

**Comment by felixfontein** [comment]
The bot reports the errors from the last Shippable run. The current one is here: https://app.shippable.com/github/ansible-collections/community.general/runs/2245/summary/console

File: plugins/inventory/proxmox.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+# Copyright (c) 2018 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: proxmox
+    plugin_type: inventory
+    short_description: proxmox inventory source
+    version_added: "2.7"
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:12]:
```suggestion
    version_added: "1.0.0"
```

File: plugins/inventory/proxmox.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        - "Uses a configuration file as an inventory source, it must end in ``.proxmox.yml`` or ``.proxmox.yaml`` and has a ``plugin: proxmox`` entry."
+        - Will retrieve the first network interface with an IP for Proxmox nodes
+        - Can retrieve LXC/QEMU configuration as facts
+    extends_documentation_fragment:
+        - inventory_cache
+    options:
+      plugin:
+        description: the name of this plugin, it should alwys be set to 'proxmox' for this plugin to recognize it as it's own.
+        required: True
+        choices: ['proxmox']
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:26]:
```suggestion
        choices: ['proxmox']
        type: str
```
Please always add types!

(The argument spec isn't yet sanity checked for plugins, so this is only reported in manual review.)

File: plugins/inventory/proxmox.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+from ansible.module_utils.six.moves.urllib.parse import urlencode
+
+
+# 3rd party imports
+try:
+    import requests
+    if LooseVersion(requests.__version__) < LooseVersion('1.1.0'):
+        raise ImportError
+except ImportError:
+    raise AnsibleError('This script requires python-requests 1.1 as a minimum version')
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:76]: Use a boolean flag for whether the import succeeded, and fail in the plugin instead.

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+# Copyright: Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: proxmox_snap
+short_description: snapshot management of instances in Proxmox VE cluster
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:11]:
```suggestion
short_description: snapshot management of instances in Proxmox VE cluster
version_added: 1.0.0
```

**Comment by Thulium-Drake** [comment]
Hi Felix,

Thanks for the reply! Well, in that case, I really don't know where I messed up, as the errors in the link you mentioned (I've been browsing it for a bit now) seem about the same as the first post the bot made with them.

However, I _did_ change the documentation to fix the issues mentioned, but the tests don't seem to agree with me.

Could you please review it for me and give me a pointer where to look?

Thanks!

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    description:
+      - Snapshot includes RAM
+    default: 'no'
+    type: bool
+  description:
+    description:
+      - Specify the description for the snapshot. Only used on the configuration web interface.
+      - This is saved as a comment inside the configuration file.
+    type: str
+    version_added: '0.2.0'
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:65]: Remove this and other `version_added` for options.

**Comment by felixfontein** [comment]
The most important sanity error is `plugins/modules/cloud/misc/proxmox_snap.py:33:9: documentation-syntax-error: DOCUMENTATION is not valid YAML`. It is probably (indirectly) causing most of the other problems.

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+      - the user to authenticate with
+    required: true
+    type: str
+  api_password:
+    description:
+      - the password to authenticate with
+      - you can use PROXMOX_PASSWORD environment variable
+    type: str
+  hostname:
+    description
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:31]:
```suggestion
    description:
```
This is probably causing the YAML parsing problem.

File: plugins/inventory/proxmox.py (hunk 2 above)
Comment (Thulium-Drake) [reply plugins/inventory/proxmox.py]: Done!

**Comment by Thulium-Drake** [comment]
ready_for_review

**Comment by Thulium-Drake** [comment]
bot_status

**Comment by felixfontein** [comment]
BTW, please note that every new plugin or module must come with tests - whether integration or unit tests (or both) depends on what's easier to do.

**Comment by Thulium-Drake** [comment]
Hi @felixfontein 

Okido, I will look into it (first time I have to make them) :-)

From what I can find in the repo, it will be unit tests, as the

**Comment by Thulium-Drake** [comment]
I will take a look at making unit tests with a friend of mine this friday (I didn't forget it, I just can't figure out how to make them :-) )

If anyone has pointers for me (there is nothing I can borrow from the existing Proxmox modules, as they also don't have tests), they are very welcome

**Comment by Thulium-Drake** [comment]
I will re-open this PR (I removed the master branch from my fork as well)

**Comment by felixfontein** [comment]
@Thulium-Drake sorry, looks like I completely forgot about this one! Please ping me once it's reopened.

**Comment by Thulium-Drake** [comment]
No rush, still need to figure out the unit test for this one.

(and it seems I broke ansible-test on my vm again)

**Pull request closed without merging**